tests/aws-cpp-sdk-s3-integration-tests/EndpointResolverIntegrationTest.cpp

```diff
@@ -0,0 +1,246 @@
+        SetHttpClientFactory(_mockClientFactory);
+    }
+
+    void TearDown() override
+    {
+        // Clean up mock clients
+        _mockClientFactory.reset();
+        _mockHttpClient.reset();
+    }
+
+    void SetupMockResponse()
+    {
+        auto mockRequest = MakeShared<Standard::StandardHttpRequest>(ALLOCATION_TAG, "mockuri", HttpMethod::HTTP_HEAD);
+        mockRequest->SetResponseStreamFactory([]() -> IOStream* {
+            return Aws::New<StringStream>(ALLOCATION_TAG, "response-string", std::ios_base::in | std::ios_base::binary);
+        });
+        auto mockResponse = MakeShared<Standard::StandardHttpResponse>(ALLOCATION_TAG, mockRequest);
+        mockResponse->SetResponseCode(HttpResponseCode::OK);
+        _mockHttpClient->AddResponseToReturn(mockResponse);
+    }
+
+    const Aws::Http::HttpRequest& ExecuteHeadBucketRequest(const S3ClientConfiguration& config)
+    {
+        SetupMockResponse();
+
+        AWSCredentials credentials{"mock", "credentials"};
+        const auto epProvider = MakeShared<S3EndpointProvider>(ALLOCATION_TAG);
+        auto testClient = MakeShared<S3TestClient>(ALLOCATION_TAG, credentials, epProvider, config);
+
+        HeadBucketRequest request;
+        request.SetBucket("test-bucket");
+
+        auto response = testClient->HeadBucket(request);
+        return _mockHttpClient->GetMostRecentHttpRequest();
+    }
+
+    std::shared_ptr<MockHttpClient> _mockHttpClient;
+    std::shared_ptr<MockHttpClientFactory> _mockClientFactory;
+};
+
+TEST_F(EndpointResolverIntegrationTest, DefaultResolution)
+{
+    S3ClientConfiguration s3Config;
+    s3Config.region = "us-east-1";
+    s3Config.profileName = "test-default";
+    s3Config.retryStrategy = MakeShared<NoRetry>(ALLOCATION_TAG);
+
+    const auto& seenRequest = ExecuteHeadBucketRequest(s3Config);
+
+    EXPECT_EQ("https://test-bucket.s3.us-east-1.amazonaws.com", seenRequest.GetUri().GetURIString());
+}
+
+TEST_F(EndpointResolverIntegrationTest, CodeProvidedEndpoint)
+{
+    S3ClientConfiguration s3Config;
+    s3Config.region = "us-east-1";
+    s3Config.endpointOverride = "https://code-provided.example.com";
+    s3Config.retryStrategy = MakeShared<NoRetry>(ALLOCATION_TAG);
+
+    const auto& seenRequest = ExecuteHeadBucketRequest(s3Config);
+
+    EXPECT_EQ("https://test-bucket.code-provided.example.com", seenRequest.GetUri().GetURIString());
+}
+
+TEST_F(EndpointResolverIntegrationTest, ServiceSpecificEnvironmentVariable)
+{
+    Aws::Environment::EnvironmentRAII setEnv{{
+        {"AWS_ENDPOINT_URL_S3", "https://custom-s3.example.com"}
+    }};
+
+    S3ClientConfiguration s3Config;
+    s3Config.region = "us-east-1";
+    s3Config.retryStrategy = MakeShared<NoRetry>(ALLOCATION_TAG);
+
+    const auto& seenRequest = ExecuteHeadBucketRequest(s3Config);
+
+    EXPECT_EQ("https://test-bucket.custom-s3.example.com", seenRequest.GetUri().GetURIString());
+}
+
+TEST_F(EndpointResolverIntegrationTest, GlobalEnvironmentVariable)
```

Comment:
I think we would want a test for `GlobalEnvironmentVariableUsedOverProfile` so that we validate that global endpoints are chosen over a specific profile endpoint.